rustls/src/msgs/deframer/handshake.rs

```diff
@@ -0,0 +1,452 @@
+use crate::msgs::codec::{u24, Codec};
+use crate::msgs::message::InboundPlainMessage;
+use crate::{ContentType, ProtocolVersion};
+
+#[derive(Debug)]
+pub(crate) struct HandshakeDeframer {
+    /// Spans covering individual handshake payloads, in order of receipt.
+    spans: Vec<FragmentSpan>,
+}
+
+impl HandshakeDeframer {
+    /// Accepts a message into the deframer.
+    ///
+    /// `containing_buffer` allows mapping the message payload to its position
+    /// in the input buffer, and thereby avoid retaining a borrow on the input
+    /// buffer.
+    ///
+    /// `outer_discard` is the rightmost extent of the original message.
+    pub(crate) fn input_message(
+        &mut self,
+        msg: InboundPlainMessage<'_>,
+        containing_buffer: &Locator,
+        outer_discard: usize,
+    ) {
+        debug_assert_eq!(msg.typ, ContentType::Handshake);
+        debug_assert!(containing_buffer.fully_contains(msg.payload));
+
+        // if our last span is incomplete, we can blindly add this as a new span --
+        // no need to attempt parsing it with `DissectHandshakeIter`.
+        //
+        // `coalesce()` will later move this new message to be contiguous with
+        // `_last_incomplete`, and reparse the result.
+        //
+        // we cannot merge these processes, because `coalesce` mutates the underlying
+        // buffer, and `msg` borrows it.
+        if let Some(_last_incomplete) = self
+            .spans
+            .last()
+            .filter(|span| !span.is_complete())
+        {
+            self.spans.push(FragmentSpan {
+                version: msg.version,
+                size: None,
+                bounds: containing_buffer.locate(msg.payload),
+                discard: outer_discard,
+            });
+            return;
+        }
+
+        // otherwise, we can expect `msg` to contain a handshake header introducing
+        // a new message (and perhaps several of them.)
+        for span in DissectHandshakeIter::new(msg, containing_buffer, outer_discard) {
+            self.spans.push(span);
+        }
+    }
+
+    /// Says what the deframer status is:
+    pub(crate) fn status(&self) -> HandshakeDeframerStatus {
+        if self.requires_coalesce().is_some() {
+            return HandshakeDeframerStatus::RequiresCoalesce;
+        }
+
+        let mut ready = 0;
+        for span in self.spans.iter() {
+            if span.is_complete() {
+                ready += 1;
+            } else {
+                return match ready {
+                    0 => HandshakeDeframerStatus::Blocked,
+                    _ => HandshakeDeframerStatus::Ready(ready),
+                };
+            }
+        }
+
+        match ready {
+            0 => HandshakeDeframerStatus::Empty,
+            _ => HandshakeDeframerStatus::Ready(ready),
+        }
+    }
+
+    pub(crate) fn is_active(&self) -> bool {
+        !matches!(self.status(), HandshakeDeframerStatus::Empty)
+    }
```

Comment:
nit: since `HandshakeDeframerStatus` impl's `PartialEq` I think this is more natural as:

```rust
    pub(crate) fn is_active(&self) -> bool {
        self.status() != HandshakeDeframerStatus::Empty
    }
```